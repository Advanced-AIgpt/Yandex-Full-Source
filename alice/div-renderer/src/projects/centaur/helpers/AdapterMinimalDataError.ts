export class AdapterMinimalDataError extends Error {
    constructor(missedData: string[], template: string) {
        super(`Missed data (${missedData.join(', ')}) in template ${template}`);
        this.name = 'AdapterMinimalDataError';
    }
}
