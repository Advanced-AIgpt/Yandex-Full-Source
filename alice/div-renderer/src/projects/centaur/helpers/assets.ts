export function getS3Asset(path: string) {
    return `https://dialogs.s3.yandex.net/smart_displays/${path}`;
}

export function getStaticS3Asset(path: string) {
    return `https://yastatic.net/s3/dialogs/smart_displays/${path}`;
}
