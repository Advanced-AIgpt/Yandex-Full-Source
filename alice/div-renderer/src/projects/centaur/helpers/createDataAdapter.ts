import Ajv, { AnySchema, JSONSchemaType } from 'ajv';
import { DeepNonNullable, DeepRequired } from 'ts-essentials';
import { IRequestState } from '../../../common/types/common';
import { ExpFlags } from '../expFlags';

export const ajv = new Ajv();

type DataAdapter<D, R> = (data: D, requestState: IRequestState) => R;

export const createDataAdapter = <D, R>(schema: AnySchema, dataAdapter: DataAdapter<D, R>) => {
    const validate = ajv.compile(schema);

    const resultDataAdapter: DataAdapter<D, R> = (data, requestState) => {
        const isDataValid = validate(data);

        if (isDataValid === false) {
            // TODO: CENTAUR-1243: log errors

            if (requestState.hasExperiment(ExpFlags.errorOnAssert)) {
                const [firstError] = validate.errors ?? [];

                throw new Error(`DataValidationError: "${firstError.instancePath}" ${firstError.message}`);
            }
        }

        return dataAdapter(data, requestState);
    };

    return resultDataAdapter;
};

export type MakeClear<T> = DeepRequired<DeepNonNullable<T>>;
export type MakeTestSchemaType<T> = JSONSchemaType<MakeClear<T>>;
