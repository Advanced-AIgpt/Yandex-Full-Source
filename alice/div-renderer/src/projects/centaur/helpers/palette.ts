export interface ColorSet {
    light: string;
    dark: string;
}

export interface ColorScheme {
    [k: string]: ColorSet | ColorScheme;
}

export type ColorNamespace<T extends ColorScheme> = {
    [K in keyof T]: T[K] extends ColorScheme ? ColorNamespace<T[K]> : string;
};

/** https://a.yandex-team.ru/arcadia/portal/morda-schema/api/search/2/paletted-color.json */
export interface PaletteColor {
    name: string;
    color: string;
}

export interface Palette {
    light: PaletteColor[];
    dark: PaletteColor[];
}

function isColorSet(x: ColorSet | ColorScheme): x is ColorSet {
    return Object.keys(x).length === 2 && 'light' in x && 'dark' in x;
}

function colorSchemeToPalette({
    colorScheme,
    scope,
}: {
    colorScheme: ColorScheme;
    scope: string;
}): Palette {
    const palette: Palette = {
        dark: [],
        light: [],
    };

    const stack: {
        path: string[];
        node: ColorScheme | ColorSet;
    }[] = Object.entries(colorScheme).map(([key, value]) => ({
        path: [key],
        node: value,
    }));

    while (stack.length) {
        const { path, node } = stack.pop() as {
            path: string[];
            node: ColorScheme | ColorSet;
        };

        if (isColorSet(node)) {
            const paletteName = `${scope}.${path.join('.')}`;
            palette.light.push({ name: paletteName, color: node.light });
            palette.dark.push({ name: paletteName, color: node.dark });
        } else {
            stack.push(
                ...Object.entries(node).map(([key, value]) => ({
                    path: [...path, key],
                    node: value,
                })),
            );
        }
    }

    return palette;
}

function colorSchemeToNamespace<T extends ColorScheme>({
    colorScheme,
    scope,
}: {
    colorScheme: T;
    scope: string;
}): ColorNamespace<T & typeof commonColorScheme> {
    const rootNamespace: ColorNamespace<ColorScheme> = {};

    const stack: {
        path: string[];
        key: string;
        node: ColorScheme | ColorSet;
        parentNamespace: ColorNamespace<any>;
    }[] = Object.entries(colorScheme).map(([key, value]) => ({
        key,
        path: [key],
        node: value,
        parentNamespace: rootNamespace,
    }));

    while (stack.length) {
        const { path, key, node, parentNamespace } = stack.pop() as {
            path: string[];
            key: string;
            node: ColorScheme | ColorSet;
            parentNamespace: ColorNamespace<any>;
        };

        if (isColorSet(node)) {
            parentNamespace[key] = `@{${scope}.${path.join('.')}}`;
        } else {
            const newNamespace: ColorNamespace<any> = {};
            parentNamespace[key] = newNamespace;

            stack.push(
                ...Object.entries(node).map(([key, value]) => ({
                    key,
                    path: [...path, key],
                    node: value,
                    parentNamespace: newNamespace,
                })),
            );
        }
    }

    return rootNamespace as ColorNamespace<T & typeof commonColorScheme>;
}

export function getPalette<T extends ColorScheme, U extends boolean | undefined = undefined>({
    colorScheme,
    scope,
    extendsCommon,
}: {
    colorScheme: T;
    scope: string;
    extendsCommon?: U;
}): {
    palette: Palette;
    ns: ColorNamespace<U extends true ? T & typeof commonColorScheme : T>;
} {
    if (extendsCommon) {
        colorScheme = {
            ...commonColorScheme,
            ...colorScheme,
        };
    }

    return {
        palette: colorSchemeToPalette({ colorScheme, scope }),
        ns: colorSchemeToNamespace({ colorScheme, scope }) as ColorNamespace<U extends true ? T & typeof commonColorScheme : T>,
    };
}

const commonColorScheme = {
    text: {
        primary: {
            light: '#000000',
            dark: '#ffffff',
        },
    },
};
