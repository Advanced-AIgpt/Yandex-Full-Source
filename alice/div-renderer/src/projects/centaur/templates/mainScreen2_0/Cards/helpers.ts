export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
