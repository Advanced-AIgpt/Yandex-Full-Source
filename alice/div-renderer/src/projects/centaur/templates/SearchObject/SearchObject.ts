import {
    ContainerBlock,
    FixedSize,
    GalleryBlock,
    ImageBlock,
    MatchParentSize,
    TemplateCard,
    Templates,
    TextBlock,
    WrapContentSize,
} from 'divcard2';
import { compact } from 'lodash';

import { NAlice } from '../../../../protos';
import { SuggestsBlock } from '../suggests';
import {
    colorLightGrey,
    colorWhiteOpacity90,
    offsetFromEdgeOfScreen,
    simpleBackground,
} from '../../style/constants';
import { sourceHost } from '../../components/sourceHost';
import { text40r, title72m } from '../../style/Text/Text';
import { MMRequest } from '../../../../common/helpers/MMRequest';
import { CloseButtonWrapper } from '../../components/CloseButtonWrapper/CloseButtonWrapper';

const firstLineHeight = title72m.line_height;

const title = (t: string) =>
    new TextBlock({
        ...title72m,
        text: t,
        width: new WrapContentSize(),
        height: new WrapContentSize(),
    });

const text = (t: string) =>
    new TextBlock({
        ...text40r,
        text_color: colorWhiteOpacity90,
        width: new WrapContentSize(),
        height: new WrapContentSize(),
        text: t,
    });

const avatar = (t: string) =>
    new ImageBlock({
        width: new FixedSize({ value: firstLineHeight }),
        height: new FixedSize({ value: firstLineHeight }),
        margins: {
            right: 32,
        },
        image_url: t,
        border: {
            corner_radius: 150,
        },
    });

const makeSearchUrl = (thumbnail: NAlice.NData.TSearchObjectData.ITGalleryImage): string =>
    (thumbnail.ThmbHref?.startsWith('/') ? 'https:' : '') +
    `${thumbnail.ThmbHref}&n=13`;

const gallery = (images: NAlice.NData.TSearchObjectData.ITGalleryImage[]) =>
    new GalleryBlock({
        width: new WrapContentSize(),
        height: new WrapContentSize(),
        item_spacing: 28,
        paddings: {
            left: offsetFromEdgeOfScreen,
            right: offsetFromEdgeOfScreen,
        },
        items: images.map(
            image =>
                new ImageBlock({
                    height: new FixedSize({ value: 366 }),
                    width: new FixedSize({ value: 366 }),
                    border: { corner_radius: 20 },
                    placeholder_color: colorLightGrey,
                    image_url: makeSearchUrl(image),
                }),
        ),
    });

function capitalizeFirstLetter(string: string): string {
    return string.charAt(0).toUpperCase() + string.slice(1);
}

export default function SearchObject(
    { Text, Title, Hostname, Image, GalleryImages }: NAlice.NData.ITSearchObjectData,
    mmRequest: MMRequest,
) {
    return new TemplateCard(new Templates({}), {
        log_id: 'search_object',
        states: [
            {
                state_id: 0,
                div: CloseButtonWrapper({
                    div: new GalleryBlock({
                        width: new WrapContentSize(),
                        height: new MatchParentSize(),
                        background: simpleBackground,
                        paddings: {
                            top: 176,
                            bottom: offsetFromEdgeOfScreen,
                        },
                        item_spacing: 0,
                        orientation: 'vertical',
                        items: compact([
                            (Image || Title) &&
                            new ContainerBlock({
                                width: new WrapContentSize(),
                                height: new WrapContentSize(),
                                content_alignment_vertical: 'top',
                                margins: {
                                    bottom: 32,
                                    left: offsetFromEdgeOfScreen,
                                    right: offsetFromEdgeOfScreen,
                                },
                                orientation: 'horizontal',
                                items: compact([Image && avatar(Image), Title && title(Title)]),
                            }),
                            Text &&
                            new ContainerBlock({
                                width: new WrapContentSize(),
                                height: new WrapContentSize(),
                                orientation: 'horizontal',
                                margins: {
                                    bottom: 16,
                                    left: offsetFromEdgeOfScreen,
                                    right: offsetFromEdgeOfScreen,
                                },
                                items: [text(capitalizeFirstLetter(Text))],
                            }),
                            Hostname && new ContainerBlock({
                                width: new WrapContentSize(),
                                height: new WrapContentSize(),
                                margins: {
                                    bottom: 40,
                                    left: offsetFromEdgeOfScreen,
                                    right: offsetFromEdgeOfScreen,
                                },
                                orientation: 'horizontal',
                                items: [sourceHost(Hostname)],
                            }),
                            GalleryImages && gallery(GalleryImages),
                            SuggestsBlock(mmRequest.ScenarioResponseBody),
                        ]),
                    }),
                }),
            },
        ],
    });
}
