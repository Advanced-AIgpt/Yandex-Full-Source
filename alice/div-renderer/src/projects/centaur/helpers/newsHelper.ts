import { getStaticS3Asset } from './assets';

const TOPICS = new Set([
    'auto',
    'business',
    'computers',
    'culture',
    'incident',
    'index',
    'koronavirus',
    'personal',
    'politics',
    'science',
    'society',
    'sport',
    'world',
]);

const ALIASES = {
    nplus1: 'science',
} as { [name: string]: string };

export function getImageByTopic(topic: string | null | undefined): string {
    let resultImage = 'news/index.png';

    if (topic) {
        if (Object.prototype.hasOwnProperty.call(ALIASES, topic)) {
            topic = ALIASES[topic];
        }

        if (TOPICS.has(topic)) {
            resultImage = `news/${topic}.png`;
        }
    }
    return getStaticS3Asset(resultImage);
}
