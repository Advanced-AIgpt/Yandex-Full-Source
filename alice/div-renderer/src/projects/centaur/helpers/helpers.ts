/* eslint-disable camelcase */
import { sortedIndexBy } from 'lodash';
import { format, isSameDay } from 'date-fns';
import { ru } from 'date-fns/locale';
import { Div, IDivData, TextBlock } from 'divcard2';
import { IRequestState } from '../../../common/types/common';
import { getTemplates, templatesRegistry } from '../../../registries/common';
import { centaurTemplatesClass } from '../index';
import { Palette } from './palette';

export { default as pluralize } from '@yandex-int/i18n/lib/plural/ru';
type TextProps = Pick<ConstructorParameters<typeof TextBlock>[0], 'text_color'|'font_size'|'font_weight'|'line_height'>;

export const textBreakpoint =
    (...breakpoints: [number, TextProps][]) =>
        (text: string) => {
            const idx = Math.min(sortedIndexBy(breakpoints, [text.length, {}], 0), breakpoints.length - 1);
            return { text, ...breakpoints[idx][1] };
        };

export const formatDate = (d: Date | string | number, f: string) => format(new Date(d), f, { locale: ru });

export const textCommand = (c: string) => `dialog://text_command?query=${encodeURI(c)}`;

export const updatedAt = (d: Date) => {
    const now = new Date();
    return `Обновлен ${isSameDay(now, d) ? 'сегодня' : formatDate(d, 'dd.MM')}`;
};

export const insertBetween = <I extends unknown, S extends unknown>(itemList: I[], separator: S) => {
    return itemList.reduce<(I | S)[]>((result, item, index) => {
        if (index !== 0) {
            result.push(separator);
        }

        result.push(item);

        return result;
    }, []);
};

export interface ITemplateCard {
    templates: {
        [name: string]: Div;
    };
    palette?: Palette;
    card: IDivData;
}

// Дива верхнего уровня. Замена для new `TemplateCard()`
export const TopLevelCard = (card: Omit<IDivData, 'variable_triggers' | 'variables'>, requestState: IRequestState, palette?: Palette) => {
    const templates = Object.assign(
        getTemplates(requestState.localTemplates, templatesRegistry),
        centaurTemplatesClass.getUsedTemplates(requestState),
        requestState.res.localTemplates.getAll(),
    );

    return {
        templates,
        palette,
        card: {
            ...card,
            variable_triggers: requestState.variableTriggers.getAll(),
            variables: [...requestState.variables],
        } as IDivData,
    };
};
