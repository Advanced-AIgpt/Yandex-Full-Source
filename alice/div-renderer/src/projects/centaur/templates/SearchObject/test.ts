import { AnonymizeDataForSnapshot } from '../../../../common/helpers/dev';
import { MMRequest } from '../../../../common/helpers/MMRequest';
import SearchObject from './SearchObject';

jest.mock('../../../../common/logger');

describe('Entity Object scenario', () => {
    it('matches snapshot', () => {
        const mmRequest = new MMRequest({}, {}, {});

        expect(
            AnonymizeDataForSnapshot(SearchObject(
                {
                    Text: 'Советский и российский государственный деятель, российский военный. Министр обороны Российской Федерации с 6 ноября 2012 года. Генерал армии. Герой Российской Федерации. Заслуженный спасатель Российской Федерации. Член Высшего совета партии «Единая Россия».',
                    Image: 'https://avatars.mds.yandex.net/get-entity_search/2048976/478239945/S120x120',
                    SearchUrl: 'https://yandex.ru/search/touch/?l10n=ru-RU&lr=213&query_source=alice&text=%D1%88%D0%BE%D0%B9%D0%B3%D1%83',
                    Url: 'http://ru.wikipedia.org/wiki/Шойгу, Сергей Кужугетович',
                    Hostname: 'http://ru.wikipedia.org',
                    Title: 'Сергей Шойгу',
                    GalleryImages: [
                        {
                            ThmbHref: '//avatars.mds.yandex.net/i?id=57077838b1e87db79dd481d75280393f-4935650-images-thumbs&ref=oo_serp',
                            ThmbHOrig: 100,
                            ThmbWOrig: 150,
                        },
                        {
                            ThmbHref: '//avatars.mds.yandex.net/i?id=107e6431697f675bd48eb4fd076a739e-5849607-images-thumbs&ref=oo_serp',
                            ThmbHOrig: 84,
                            ThmbWOrig: 150,
                        },
                        {
                            ThmbHref: '//avatars.mds.yandex.net/i?id=551fda7fd5dae006024e01de3c45abc3-5827319-images-thumbs&ref=oo_serp',
                            ThmbHOrig: 84,
                            ThmbWOrig: 150,
                        },
                        {
                            ThmbHref: '//avatars.mds.yandex.net/i?id=f264702fa8f6e5bc4402a6a0e3801e8a-4034704-images-thumbs&ref=oo_serp',
                            ThmbHOrig: 96,
                            ThmbWOrig: 150,
                        },
                        {
                            ThmbHref: '//avatars.mds.yandex.net/i?id=a9ab9de1ff86c2443ae52a6f87e0bbc9-5905309-images-thumbs&ref=oo_serp',
                            ThmbHOrig: 92,
                            ThmbWOrig: 150,
                        },
                        {
                            ThmbHref: '//avatars.mds.yandex.net/i?id=f93726267a6c2762c0e45910d79e7a98-5259114-images-thumbs&ref=oo_serp',
                            ThmbHOrig: 92,
                            ThmbWOrig: 150,
                        },
                        {
                            ThmbHref: '//avatars.mds.yandex.net/i?id=9cba55dc901856709675cfd321f962ec-5896417-images-thumbs&ref=oo_serp',
                            ThmbHOrig: 100,
                            ThmbWOrig: 150,
                        },
                        {
                            ThmbHref: '//avatars.mds.yandex.net/i?id=6acac7464255affa370b67e7d7e0fe25-5485004-images-thumbs&ref=oo_serp',
                            ThmbHOrig: 84,
                            ThmbWOrig: 150,
                        },
                    ],
                },
                mmRequest,
            )),
        ).toMatchSnapshot();
    });
});
