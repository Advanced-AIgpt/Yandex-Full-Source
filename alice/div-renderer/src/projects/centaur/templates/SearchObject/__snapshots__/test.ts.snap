// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity Object scenario matches snapshot 1`] = `
Object {
  "card": Object {
    "log_id": "search_object",
    "states": Array [
      Object {
        "div": Object {
          "height": Object {
            "type": "match_parent",
            "weight": 1,
          },
          "items": Array [
            Object {
              "background": Array [
                Object {
                  "color": "#121316",
                  "type": "solid",
                },
              ],
              "height": Object {
                "type": "match_parent",
              },
              "item_spacing": 0,
              "items": Array [
                Object {
                  "content_alignment_vertical": "top",
                  "height": Object {
                    "type": "wrap_content",
                  },
                  "items": Array [
                    Object {
                      "border": Object {
                        "corner_radius": 150,
                      },
                      "height": Object {
                        "type": "fixed",
                        "value": 88,
                      },
                      "image_url": "https://avatars.mds.yandex.net/get-entity_search/2048976/478239945/S120x120",
                      "margins": Object {
                        "right": 32,
                      },
                      "type": "image",
                      "width": Object {
                        "type": "fixed",
                        "value": 88,
                      },
                    },
                    Object {
                      "font_size": 72,
                      "font_weight": "medium",
                      "height": Object {
                        "type": "wrap_content",
                      },
                      "line_height": 88,
                      "text": "Сергей Шойгу",
                      "text_color": "#fff",
                      "type": "text",
                      "width": Object {
                        "type": "wrap_content",
                      },
                    },
                  ],
                  "margins": Object {
                    "bottom": 32,
                    "left": 48,
                    "right": 48,
                  },
                  "orientation": "horizontal",
                  "type": "container",
                  "width": Object {
                    "type": "wrap_content",
                  },
                },
                Object {
                  "height": Object {
                    "type": "wrap_content",
                  },
                  "items": Array [
                    Object {
                      "font_size": 40,
                      "font_weight": "regular",
                      "height": Object {
                        "type": "wrap_content",
                      },
                      "line_height": 56,
                      "text": "Советский и российский государственный деятель, российский военный. Министр обороны Российской Федерации с 6 ноября 2012 года. Генерал армии. Герой Российской Федерации. Заслуженный спасатель Российской Федерации. Член Высшего совета партии «Единая Россия».",
                      "text_color": "#e6ffffff",
                      "type": "text",
                      "width": Object {
                        "type": "wrap_content",
                      },
                    },
                  ],
                  "margins": Object {
                    "bottom": 16,
                    "left": 48,
                    "right": 48,
                  },
                  "orientation": "horizontal",
                  "type": "container",
                  "width": Object {
                    "type": "wrap_content",
                  },
                },
                Object {
                  "height": Object {
                    "type": "wrap_content",
                  },
                  "items": Array [
                    Object {
                      "font_size": 32,
                      "font_weight": "medium",
                      "height": Object {
                        "type": "wrap_content",
                      },
                      "line_height": 40,
                      "text": "ru.wikipedia.org",
                      "text_color": "#80ffffff",
                      "type": "text",
                      "width": Object {
                        "type": "wrap_content",
                      },
                    },
                  ],
                  "margins": Object {
                    "bottom": 40,
                    "left": 48,
                    "right": 48,
                  },
                  "orientation": "horizontal",
                  "type": "container",
                  "width": Object {
                    "type": "wrap_content",
                  },
                },
                Object {
                  "height": Object {
                    "type": "wrap_content",
                  },
                  "item_spacing": 28,
                  "items": Array [
                    Object {
                      "border": Object {
                        "corner_radius": 20,
                      },
                      "height": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                      "image_url": "https://avatars.mds.yandex.net/i?id=57077838b1e87db79dd481d75280393f-4935650-images-thumbs&ref=oo_serp&n=13",
                      "placeholder_color": "#ccc",
                      "type": "image",
                      "width": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                    },
                    Object {
                      "border": Object {
                        "corner_radius": 20,
                      },
                      "height": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                      "image_url": "https://avatars.mds.yandex.net/i?id=107e6431697f675bd48eb4fd076a739e-5849607-images-thumbs&ref=oo_serp&n=13",
                      "placeholder_color": "#ccc",
                      "type": "image",
                      "width": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                    },
                    Object {
                      "border": Object {
                        "corner_radius": 20,
                      },
                      "height": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                      "image_url": "https://avatars.mds.yandex.net/i?id=551fda7fd5dae006024e01de3c45abc3-5827319-images-thumbs&ref=oo_serp&n=13",
                      "placeholder_color": "#ccc",
                      "type": "image",
                      "width": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                    },
                    Object {
                      "border": Object {
                        "corner_radius": 20,
                      },
                      "height": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                      "image_url": "https://avatars.mds.yandex.net/i?id=f264702fa8f6e5bc4402a6a0e3801e8a-4034704-images-thumbs&ref=oo_serp&n=13",
                      "placeholder_color": "#ccc",
                      "type": "image",
                      "width": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                    },
                    Object {
                      "border": Object {
                        "corner_radius": 20,
                      },
                      "height": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                      "image_url": "https://avatars.mds.yandex.net/i?id=a9ab9de1ff86c2443ae52a6f87e0bbc9-5905309-images-thumbs&ref=oo_serp&n=13",
                      "placeholder_color": "#ccc",
                      "type": "image",
                      "width": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                    },
                    Object {
                      "border": Object {
                        "corner_radius": 20,
                      },
                      "height": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                      "image_url": "https://avatars.mds.yandex.net/i?id=f93726267a6c2762c0e45910d79e7a98-5259114-images-thumbs&ref=oo_serp&n=13",
                      "placeholder_color": "#ccc",
                      "type": "image",
                      "width": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                    },
                    Object {
                      "border": Object {
                        "corner_radius": 20,
                      },
                      "height": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                      "image_url": "https://avatars.mds.yandex.net/i?id=9cba55dc901856709675cfd321f962ec-5896417-images-thumbs&ref=oo_serp&n=13",
                      "placeholder_color": "#ccc",
                      "type": "image",
                      "width": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                    },
                    Object {
                      "border": Object {
                        "corner_radius": 20,
                      },
                      "height": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                      "image_url": "https://avatars.mds.yandex.net/i?id=6acac7464255affa370b67e7d7e0fe25-5485004-images-thumbs&ref=oo_serp&n=13",
                      "placeholder_color": "#ccc",
                      "type": "image",
                      "width": Object {
                        "type": "fixed",
                        "value": 366,
                      },
                    },
                  ],
                  "paddings": Object {
                    "left": 48,
                    "right": 48,
                  },
                  "type": "gallery",
                  "width": Object {
                    "type": "wrap_content",
                  },
                },
                Object {
                  "item_spacing": 24,
                  "items": Array [],
                  "type": "gallery",
                },
              ],
              "orientation": "vertical",
              "paddings": Object {
                "bottom": 48,
                "top": 176,
              },
              "type": "gallery",
              "width": Object {
                "type": "wrap_content",
              },
            },
            Object {
              "alignment_horizontal": "right",
              "alignment_vertical": "top",
              "items": Array [
                Object {
                  "actions": Array [
                    Object {
                      "log_id": "close_fullscreen",
                      "url": "centaur://local_command?local_commands=[%7B%22command%22%3A%22close_layer%22%2C%22layer%22%3A%22DIALOG%22%2C%22stop_conversation%22%3Atrue%7D]",
                    },
                  ],
                  "alignment_horizontal": "right",
                  "background": Array [
                    Object {
                      "color": "#1affffff",
                      "type": "solid",
                    },
                  ],
                  "border": Object {
                    "corner_radius": 20,
                  },
                  "height": Object {
                    "type": "fixed",
                    "value": 88,
                  },
                  "image_url": "https://yastatic.net/s3/dialogs/smart_displays/icons/close_wt_border.png",
                  "margins": Object {
                    "right": 36,
                    "top": 36,
                  },
                  "paddings": Object {
                    "bottom": 22,
                    "left": 22,
                    "right": 22,
                    "top": 22,
                  },
                  "preload_required": 1,
                  "scale": "fill",
                  "type": "image",
                  "width": Object {
                    "type": "fixed",
                    "value": 88,
                  },
                },
              ],
              "type": "container",
            },
          ],
          "orientation": "overlap",
          "type": "container",
          "width": Object {
            "type": "match_parent",
            "weight": 1,
          },
        },
        "state_id": 0,
      },
    ],
  },
  "templates": Object {},
}
`;
